http/cookies.lua

```diff
@@ -0,0 +1,353 @@
+		end
+	else
+		cookie.expires = math.huge
+		table.insert(cookies, 1, cookie)
+	end
+
+	local by_domain = cookies[domain]
+	if not by_domain then
+		by_domain = {}
+		cookies[domain] = by_domain
+	end
+	local by_path = by_domain[path]
+	if not by_path then
+		by_path = {}
+		by_domain[path] = by_path
+	end
+	by_path[key] = cookie
+	return true
+end
+
+function cookiejar_methods:get(domain, path, key)
+	local by_domain = self.cookies[domain]
+	if not by_domain then
+		return nil
+	end
+	local by_path = by_domain[path]
+	if not by_path then
+		return nil
+	end
+	return by_path[key]
+end
+
+function cookiejar_methods:remove_cookie(cookie)
+	local cookies = self.cookies
+	for i=1, #cookies do
+		if cookie == cookies[i] then
+			table.remove(cookies, i)
+			cookies[cookie.domain][cookie.path][cookie.key] = nil
+			return true
+		end
+	end
+	return false
+end
+
+function cookiejar_methods:remove_cookies(cookies)
+	local cookie_hashes = {}
+	for _, key in pairs(cookies) do
+		cookie_hashes[key] = true
+	end
+	local s_cookies = self.cookies
+	local n = #s_cookies
+	local start_hole = 0
+	for i=1, n do
+		local value = s_cookies[i]
+		if value and cookie_hashes[value] then
+			s_cookies[i] = nil
+			local by_domain = s_cookies[value.domain]
+			local by_path = by_domain[value.path]
+			by_path[value.key] = nil
+			if not next(by_path) then
+				by_domain[value.path] = nil
+				if not next(by_domain) then
+					s_cookies[value.domain] = nil
+				end
+			end
+			if start_hole == 0 then
+				-- start_hole hasn't been initialized, and a hole exists, so
+				-- start the hole at the current position
+				start_hole = i
+			end
+		elseif start_hole ~= 0 then
+			-- a cookie exists and isn't being removed, so shift the cookie
+			-- downwards to the start of the hole
+			s_cookies[start_hole] = value
+			s_cookies[i] = nil
+			start_hole = start_hole + 1
+		end
+	end
+end
+
+local function get_expired(jar, time)
+	time = time or os.time()
+	local cookies = jar.cookies
+	local returned_cookies = {}
+	for i=#cookies, 1, -1 do
+		local cookie = cookies[i]
+		if cookie.expires > time then
+			break
+		end
+		returned_cookies[#returned_cookies + 1] = cookie
+	end
+	return returned_cookies
+end
+
+function cookiejar_methods:remove_expired(time)
+	self:remove_cookies(get_expired(self, time))
+end
+
+function cookiejar_methods:trim(size)
+	self:remove_expired()
+	local cookies = self.cookies
+	if #cookies > size then
+		for i=#cookies, size + 1, -1 do
+			local cookie = cookies[i]
+			cookies[i] = nil
+			local by_domain = cookies[cookie.domain]
+			local by_path = by_domain[cookie.path]
+			by_path[cookie.key] = nil
+			if not next(by_path) then
+				by_domain[cookie.path] = nil
+				if not next(by_domain) then
+					cookies[cookie.domain] = nil
+				end
+			end
+		end
+	end
+end
+
+local function serialize_cookies(cookies)
+	local out_values = {}
+	for _, cookie in pairs(cookies) do
+		out_values[#out_values + 1] = cookie.key .. "=" .. cookie.value
```

Comment:
*TODO:* Use a value `n` to avoid getting length every time, then increment every time